### PR TITLE
Replace docker/pkg/fileutils with github.com/moby/patternmatcher

### PR DESCRIPTION
**Pull request by thaJeztah**
- replaces https://github.com/tonistiigi/fsutil/pull/105
- closes https://github.com/tonistiigi/fsutil/pull/105
- relates to https://github.com/moby/buildkit/pull/2150#issuecomment-857253756
- relates to https://github.com/moby/moby/pull/44226

The file matching functions were extracted to https://github.com/moby/patternmatcher

Also updating docker to the latest 20.10 patch release in the bench module, and updating the go.mod syntax to 1.18

### CODE REVIEW

**Comment by thaJeztah** [comment]
@tonistiigi @aaronlehmann @cpuguy83 PTAL (I haven't tagged a release yet, but can do if we're all happy with the current state)

**Review by tonistiigi — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]

**Comment by thaJeztah** [comment]
hm.. something is failing; let me check

**Comment by thaJeztah** [comment]
Ah; linter is still running go 1.17; need to update github actions to go 1.18/1.19; let me open a PR

```
 go mod tidy: go.mod file indicates go 1.18, but maximum supported version is 1.17
```

**Comment by thaJeztah** [comment]
opened https://github.com/tonistiigi/fsutil/pull/137

**Review by aaronlehmann — APPROVED** [review]

**Comment by aaronlehmann** [comment]
Thanks for doing this!

**Comment by thaJeztah** [comment]
It kept dropping off my list to do. Not _entirely_ out of the woods yet; see https://github.com/moby/moby/pull/44227, but we can temporarily depend on the aliases I created in docker/docker for that (until the buildkit version we use there is compatible with the latest tonistiigi/fsutil.

**Comment by thaJeztah** [comment]
Ah, dang... and more to do; looks like we're still using `io/ioutil` in this repo;

```
#26 9.826 fs.go:6:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
#26 9.826 	"io/ioutil"
#26 9.826 	^
```

File: copy/copy_test.go [hunk 1]
```diff
@@ -137,39 +127,31 @@ func TestCopyDevicesAndFifo(t *testing.T) {
...
+	err := Copy(context.TODO(), t1, "foo.txt", t2, "/")
 	require.NoError(t, err)
 
 	err = fstest.CheckDirectoryEqual(t1, t2)
 	require.NoError(t, err)
 
-	t3, err := ioutil.TempDir("", "test")
-	require.NoError(t, err)
-	defer os.RemoveAll(t2)
+	t3 := t.TempDir()
```
Comment (thaJeztah) [inline copy/copy_test.go:146]: I saw a couple of these, and I guess those were copy/paste errors; we create `t3`, but add a defer to remove `t2` (I guess CI will show if it was not intentional 😂)

**Comment by thaJeztah** [comment]
Nice; all green now; I'll tag v0.5.0, and update this PR 👍

**Comment by thaJeztah** [comment]
done; should all be good now

**Pull request merged**